blocklylib-core/src/main/java/com/google/blockly/model/BlockTemplate.java

```diff
@@ -0,0 +1,513 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.blockly.model;
+
+import android.text.TextUtils;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Template of a block, describing the initial state of a new block.
+ *
+ * The API of this class is designed to be used in a literate manner, as an input to
+ * {@link BlockFactory#obtain}. The static method {@link BlockFactory#block()} (which returns a
+ * {@code BlockTemplate}) can be used as a via static import to read like English.
+ *
+ * <pre>
+ * {@code factory.obtain(block().shadow().ofType("math_number").atPosition(25,56));}
+ * </pre>
+ */
+public class BlockTemplate {
```

Comment:
This is mixing a few use cases and is a bit of a dangerous API.
1. Quickly applying a preconfigured state to several blocks.
2. Quickly setting up a block with some simple state.
3. Storing all the data for a block loaded from XML

1 and 2 work well for a public API, but 3 includes state that can't be applied to more than one block (input and next children). I think you should consider reducing the scope of this class to just 1 & 2 and then having a private extension that adds the non-replicable state for use by the XML parser.